### PR TITLE
Document and refactor the CanonicalIdResolver extension point

### PR DESCRIPTION
**Pull request by oleg-nenashev**
The extension point has been introduced by @ndeloof  in https://github.com/jenkinsci/jenkins/commit/af42534defea4503347dfc0ec442fadf51cf3336
It does not follow common extension point practices (e.g. “all()” methods) and has no Javadoc.

### Proposed changelog entries

* Developer: Add documentation and convenience methods to the `CanonicalIdResolver` extension point
  * Extension point docs: https://jenkins.io/doc/developer/extensions/jenkins-core/#user-wbr-canonicalidresolver

No functional code changes, so no JIRA ticket/tests. I can create them on-demand

### Desired reviewers

@reviewbybees @ndeloof

### CODE REVIEW

File: core/src/main/java/hudson/model/User.java [hunk 1]
```diff
@@ -401,23 +401,16 @@ public synchronized void doSubmitDescription( StaplerRequest req, StaplerRespons
      *      An existing or created user. May be {@code null} if a user does not exist and
      *      {@code create} is false.
      */
-    public static @Nullable User get(String idOrFullName, boolean create, Map context) {
+    public static @Nullable User get(String idOrFullName, boolean create, @Nonnull Map context) {
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/User.java]: I cannot find usages of this parameter at all, but I decided to be on the safe side

**Comment by oleg-nenashev** [comment]
@reviewbybees ping

**Review by ndeloof — APPROVED** [review]

**Comment by oleg-nenashev** [comment]
Thanks @ndeloof! Just doing some polishing while I am around

**Comment by oleg-nenashev** [comment]
@reviewbybees done

**Pull request merged**